client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,159 @@
+	getEmailForwards,
+	isAddingEmailForward,
+} from 'calypso/state/selectors/get-email-forwards';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( this.props.isAddingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+	};
+
+	onForwardAdd = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
+
+	renderActionsButtons( addMoreButton = false ) {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				{ addMoreButton && (
+					<Button
+						className="email-forwarding-add-new-compact-list__add-another-forward-button"
+						onClick={ this.onForwardAdd }
+					>
+						<Gridicon icon="plus" />
+						<span>{ translate( 'Add another forward' ) }</span>
+					</Button>
+				) }
+
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.hasValidEmailForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
+		emailForwards.splice( index, 1 );
+		this.setState( { emailForwards } );
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		const array = this.state.emailForwards;
+		array[ index ][ name ] = value;
+
+		const validEmailForward = validateAllFields( array[ index ] );
+		array[ index ].isValid =
+			validEmailForward.mailbox.length === 0 && validEmailForward.destination.length === 0;
+
+		this.setState( { emailForwards: array } );
+	};
+
+	render() {
+		const { emailForwardSuccess, selectedDomainName, selectedSiteSlug } = this.props;
+
+		if ( emailForwardSuccess ) {
+			page( emailManagement( selectedSiteSlug, selectedDomainName ) );
+		}
```

Comment:
Please `return` after this line -- this may be triggering the weird issues 'Seni reported.

Actually, I am also worried that this is the wrong place for the redirect logic. This component could be added anywhere in the UI, so redirecting to the email management page doesn't belong in this component.

I think it would be better to set up an optional `onSuccessRedirect` prop to handle this better.

```suggestion
		// We could also use something like `onSuccessRedirectHandler` as a function, but I don't know if it's worth generalising that much
		if ( emailForwardSuccess && onSuccessRedirectDestination ) {
			page( onSuccessRedirectDestination );
			return;
		}
```

Reply:
I've also took advice from react itself, it said in the console, better to move that code to componentDidUpdate